src/Servers/Kestrel/samples/SampleApp/ClientCertBufferingFeature.cs

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Security.Cryptography.X509Certificates;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Connections.Features;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.AspNetCore.WebUtilities;
+
+namespace SampleApp
+{
+    internal static class ClientCertBufferingExtensions
+    {
+        // Buffers HTTP/1.x request bodies received over TLS (https) if a client certificate needs to be negotiated.
+        // This avoids the issue where POST data is received during the certificate negotiation:
+        // InvalidOperationException: Received data during renegotiation.
+        public static IApplicationBuilder UseClientCertBuffering(this IApplicationBuilder builder)
+        {
+            return builder.Use((context, next) =>
+            {
+                var tlsFeature = context.Features.Get<ITlsConnectionFeature>();
+                var bodyFeature = context.Features.Get<IHttpRequestBodyDetectionFeature>();
+                var connectionItems = context.Features.Get<IConnectionItemsFeature>();
+
+                // Look for TLS connections that don't already have a client cert, and requests that could have a body.
+                if (tlsFeature != null && tlsFeature.ClientCertificate == null && bodyFeature.CanHaveBody
+                    && !connectionItems.Items.TryGetValue("tls.clientcert.negotiated", out var _))
```

Comment:
Why not?

```suggestion
                    && (tlsFeature is not ClientCertBufferingFeature))
```

Reply:
ClientCertBufferingFeature gets stored in the HttpContext's FeatureCollection, not the connection's FeatureCollection. It's reverted before the next request. Items persist across requests.